rpc/src/v1/helpers/light_fetch.rs

```diff
@@ -299,78 +301,67 @@ impl LightFetch {
-			Some(from) => (from.number(), from),
-			None => return Either::A(future::err(errors::unknown_block())),
-		};
+		const MAX_BLOCK_RANGE: u64 = 1000;
 
-		match block_number(filter.to_block) {
-			Some(to) if to < from_block_number || from_block_number > best_number 
-				=> return Either::A(future::ok(Vec::new())),
-			Some(_) => (),
-			_ => return Either::A(future::err(errors::unknown_block())),
-		}
-
-		let maybe_future = self.sync.with_context(move |ctx| {
-			// find all headers which match the filter, and fetch the receipts for each one.
-			// match them with their numbers for easy sorting later.
-			let bit_combos = filter.bloom_possibilities();
-			let receipts_futures: Vec<_> = self.client.ancestry_iter(filter.to_block)
-				.take_while(|ref hdr| hdr.number() != from_block_number)
-				.chain(Some(from_block_header))
-				.filter(|ref hdr| {
-					let hdr_bloom = hdr.log_bloom();
-					bit_combos.iter().any(|bloom| hdr_bloom.contains_bloom(bloom))
-				})
-				.map(|hdr| (hdr.number(), hdr.hash(), request::BlockReceipts(hdr.into())))
-				.map(|(num, hash, req)| self.on_demand.request(ctx, req).expect(NO_INVALID_BACK_REFS).map(move |x| (num, hash, x)))
-				.collect();
+		let fetcher = self.clone();
+		self.headers_range_by_block_id(filter.from_block, filter.to_block, MAX_BLOCK_RANGE)
+			.and_then(move |mut headers| {
+				if headers.is_empty() {
+					return Either::A(future::ok(Vec::new()));
+				}
 
-			// as the receipts come in, find logs within them which match the filter.
-			// insert them into a BTreeMap to maintain order by number and block index.
-			stream::futures_unordered(receipts_futures)
-				.fold(BTreeMap::new(), move |mut matches, (num, hash, receipts)| {
-					let mut block_index = 0;
-					for (transaction_index, receipt) in receipts.into_iter().enumerate() {
-						for (transaction_log_index, log) in receipt.logs.into_iter().enumerate() {
-							if filter.matches(&log) {
-								matches.insert((num, block_index), Log {
-									address: log.address.into(),
-									topics: log.topics.into_iter().map(Into::into).collect(),
-									data: log.data.into(),
-									block_hash: Some(hash.into()),
-									block_number: Some(num.into()),
-									// No way to easily retrieve transaction hash, so let's just skip it.
-									transaction_hash: None,
-									transaction_index: Some(transaction_index.into()),
-									log_index: Some(block_index.into()),
-									transaction_log_index: Some(transaction_log_index.into()),
-									log_type: "mined".into(),
-									removed: false,
-								});
+				let on_demand = &fetcher.on_demand;
+
+				let maybe_future = fetcher.sync.with_context(move |ctx| {
+					// find all headers which match the filter, and fetch the receipts for each one.
+					// match them with their numbers for easy sorting later.
+					let bit_combos = filter.bloom_possibilities();
+					let receipts_futures: Vec<_> = headers.drain(..)
```

Comment:
A minor thing, isn't better to use `into_iter` instead of `drain` in this context because we are not reusing the `headers` after the operation?

(`drain` is borrowing the vector (ends up with an empty vector) vs `into_iter` is taking ownership)